src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestState.cs

```diff
@@ -123,6 +123,22 @@ public void ClearSendRequestState()
         public long? ExpectedBytesToRead { get; set; }
         public long CurrentBytesRead { get; set; }
 
+        // TODO (Issue 2505): temporary pinned buffer caches of 1 item. Will be replaced by PinnableBufferCache.
+        private GCHandle _cachedReceivePinnedBuffer = default(GCHandle);
+
+        public void PinReceiveBuffer(byte[] buffer)
+        {
+            if (!_cachedReceivePinnedBuffer.IsAllocated || _cachedReceivePinnedBuffer.Target != buffer)
```

Comment:
The WinHttpRequestState object relies on the caller (and there is only one caller) to assert that.  In particular, the only async operation possible on this pinned receive buffer is from WinHttpResponseStream.ReadAsync().  There is already a check in that public API surface to reject calls to it if the previous ReadAsync is in progress.